### PR TITLE
flash: redirect gd32vf103 driver to stm32f1x

### PR DESCRIPTION
**Pull request by fanghuaqi**
Hi @timsifive , I modify the flash driver like this, is it ok?

Most of the code of stm32f1x are reused, I just add struct of gd32vf103, so `flash gd32vf103` can be still recognized.

### CODE REVIEW

**Review by timsifive — CHANGES_REQUESTED** [review]
This looks like a good framework.

You should make a gd32vf103_flash_bank_command that prints an error, and then calls stm32x_flash_bank_command, that way at least users will be notified.

I propose we then fully remove this temporary solution in June of 2023.

**Comment by fanghuaqi** [comment]
Hi @timsifive , I have add a `gd32vf103_flash_bank_command` function to print warning message to tell user to use stm32f1x, I want to call `stm32x_flash_bank_command` this function in the `gd32vf103_flash_bank_command` implementation, but didn't find a way, so I just copy the code from `stm32x_flash_bank_command`.

File: bootstrap [hunk 1]
```diff
@@ -1,4 +1,4 @@
-#!/bin/sh
+#!/bin/env sh
```
Comment (timsifive) [inline bootstrap]: This has nothing to do with this change, and doesn't have anything to do with RISC-V either. Can you submit this upstream instead?

File: src/flash/nor/stm32f1x.c [hunk 2]
```diff
@@ -1751,3 +1751,54 @@ const struct flash_driver stm32f1x_flash = {
...
 	.free_driver_priv = default_flash_free_driver_priv,
 };
+
+/* flash bank gd32vf103 <base> <size> 0 0 <target#>
+ */
+FLASH_BANK_COMMAND_HANDLER(gd32vf103_flash_bank_command)
+{
+	struct stm32x_flash_bank *stm32x_info;
+
+	LOG_WARNING("The gd32vf103 might be removed in future, please use stm32f1x instead.");
```
Comment (timsifive) [inline src/flash/nor/stm32f1x.c]:
```suggestion
	LOG_WARNING("The gd32vf103 flash target will be removed in June of 2023, please use stm32f1x instead.");
```

File: src/flash/nor/stm32f1x.c [hunk 3]
```diff
@@ -1751,3 +1751,54 @@ const struct flash_driver stm32f1x_flash = {
...
+	if (CMD_ARGC < 6)
```
Comment (timsifive) [inline src/flash/nor/stm32f1x.c:1763]: Can you add a comment that the rest of this function was copied from stm32x...?

**Review by timsifive — CHANGES_REQUESTED** [review]

File: bootstrap (hunk 1 above)
Comment (fanghuaqi) [reply bootstrap]: Sorry, this is mistaken added. I will remove it.

File: src/flash/nor/stm32f1x.c (hunk 2 above)
Comment (fanghuaqi) [reply src/flash/nor/stm32f1x.c]: Done

File: src/flash/nor/stm32f1x.c (hunk 3 above)
Comment (fanghuaqi) [reply src/flash/nor/stm32f1x.c:1763]: Comment message added in source code.

**Comment by fanghuaqi** [comment]
Thanks @timsifive for such detailed review, I forced pushed the commit, could you help to review it again.

**Review by timsifive — DISMISSED** [review]
Thanks! I'll merge this tomorrow unless somebody else comments.

File: src/flash/nor/stm32f1x.c [hunk 4]
```diff
@@ -1751,3 +1751,55 @@ const struct flash_driver stm32f1x_flash = {
...
+	/* The flash write must be aligned to a halfword boundary */
+	bank->write_start_alignment = bank->write_end_alignment = 2;
+	return ERROR_OK;
+}
+
+static const struct command_registration gd32vf103_command_handlers[] = {
+	{
+		.name = "gd32vf103",
+		.mode = COMMAND_ANY,
+		.help = "gd32vf103 flash command group(reuse stm32f1x)",
```
Comment (JanMatCodasip) [inline src/flash/nor/stm32f1x.c]: Suggestion for minor language correction:

```suggestion
		.help = "gd32vf103 flash command group (identical to flash commands from stm32f1x)",
```

File: src/flash/nor/stm32f1x.c [hunk 5]
```diff
@@ -1751,3 +1751,55 @@ const struct flash_driver stm32f1x_flash = {
...
 	.free_driver_priv = default_flash_free_driver_priv,
 };
+
+/* flash bank gd32vf103 <base> <size> 0 0 <target#>
+ */
+FLASH_BANK_COMMAND_HANDLER(gd32vf103_flash_bank_command)
+{
+	struct stm32x_flash_bank *stm32x_info;
+
+	LOG_WARNING("The gd32vf103 flash target will be removed in June of 2023, please use stm32f1x instead.");
```
Comment (JanMatCodasip) [inline src/flash/nor/stm32f1x.c]: Here I'd suggest a minor change to improve the visibility of the message.

The word `DEPRECATED` is commonly used in such messages other parts of OpenOCD, too.

```suggestion
	LOG_WARNING("DEPRECATED: The gd32vf103 flash target will be removed in June of 2023, please use stm32f1x instead.");
```

**Review by JanMatCodasip — COMMENTED** [review]
I have posted two minor suggestions for the log messages.

File: src/flash/nor/stm32f1x.c (hunk 5 above)
Comment (fanghuaqi) [reply src/flash/nor/stm32f1x.c]: Fixed

File: src/flash/nor/stm32f1x.c (hunk 4 above)
Comment (fanghuaqi) [reply src/flash/nor/stm32f1x.c]: Fixed.

**Comment by fanghuaqi** [comment]
Forced pushed changes requested by @JanMatCodasip

**Review by JanMatCodasip — APPROVED** [review]
Thanks, looks good.

**Pull request merged**